### ISSUE TITLE
Incorrect behavior with numpy structured array

### ISSUE THREAD

**Issue by Anthony-C31** [numba/numba#7859]
Hello,
I have a different behavior when executing the following code with or with numba:

import numpy as np
import numba as nb
table = np.zeros(2, dtype=([('field1', np.uint8), ('field2', '<U1')]))
table[0] = (1, 'A')
table[1] = (2, 'B')
@nb.njit
def foo(table):
    print(table['field1'] == 1)
    print(table['field2'] == 'A')
    return
foo(table)

Without numba: (prints two vectors of boolean, OK)
[ True False]
[ True False]

With numba: (second print return just a boolean, NOK)
[ True False]
False

Thanks for your help!
Anthony

**Comment by sklam** [comment]
Bug confirmed. The problem is in the second `==`. It's currently typed as `(unaligned array([unichr x 1], 1d, A), unicode_type) -> Literal[bool](False)`. Need to check which definition it has selected.

**Comment by sklam** [comment]
Debug summary:

- Current problem occurs because the following unicode equality is taken: https://github.com/numba/numba/blob/53ea89fee1a19471ce1b1907f9f3d98acd1b6730/numba/cpython/unicode.py#L482-L485  This definition is hard-wired to `False` because it thinks it is comparing unicode to non-unicode.
- What should happen is the Array `__eq__` should have recognized this.
- Related problem `np.asarray("A")` is not supported.

Minimal reproducer:

```python
import numpy as np
import numba as nb


@nb.njit
def foo(a):
    return a == 'A'

a = np.array(['A', 'B'], dtype=np.dtype('<U1'))
x = foo(a)
y = foo.py_func(a)
np.testing.assert_equal(x, y)
  ```

**Comment by dlee992** [comment]
I would like to contribute a bug fix for this issue, if u don't begin to code about this. I have some experience about Numba source code when I develop SDC. @gmarkall @sklam

**Comment by dlee992** [comment]
Hi, @sklam @stuartarchibald . I am trying to fix this issue. Right now, I found two string-related types, `CharSeq` and `UnicodeCharSeq` defined for numpy array, and `Unicode` type defined in a separate file. Does Numba have any method to convert_to/from between these types? e.g., convert `Unicode` to `UnicodeCharSeq`.

**Comment by sklam** [comment]
@dlee992, yes, i agree adding `equal` ufunc for strings and unicode may fix this issue. I am guessing bulk of the work is extending `ufunc_find_matching_loop` for `SS->?`/`UU->?`. For writing the equivalent to `int_eq_impl` for unicode, take a look at `complex_div_impl`: https://github.com/numba/numba/blob/e5bbae7de22fa8c0b23dcfc8c9a946e5d37ee509/numba/cpython/numbers.py#L1060-L1089

See how it trigger the compiler within itself. The unicode equal ufunc can likely reuse the unicode equality like this: 

```python
def unicode_eq_impl(context, builder, sig, args):
    def eq(a, b):
        return a == b

    res = context.compile_internal(builder, eq, sig, args)
    return impl_ret_untracked(context, builder, sig.return_type, res)
```

**Comment by sklam** [comment]
Looks like numpy is not using the ufunc for unicode/string array equality.  

```python
In [3]: a=np.asarray("ABC")

In [4]: a
Out[4]: array('ABC', dtype='<U3')

In [5]: np.equal(a, a)
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
<ipython-input-5-e945fc627bdf> in <module>
----> 1 np.equal(a, a)

UFuncTypeError: ufunc 'equal' did not contain a loop with signature matching types (dtype('<U3'), dtype('<U3')) -> dtype('bool')
```

It's not obvious from a quick search where the numpy code for `a == a` is.

**Comment by stuartarchibald** [comment]
> It's not obvious from a quick search where the numpy code for `a == a` is.

It's probably this:
https://github.com/numpy/numpy/blob/26c1e04b3af572ed4ee1cc2f9ac3faca27858dc5/numpy/core/src/multiarray/arrayobject.c#L1275
it's in the rich comparison slot for the array type:
https://github.com/numpy/numpy/blob/26c1e04b3af572ed4ee1cc2f9ac3faca27858dc5/numpy/core/src/multiarray/arrayobject.c#L1709

Note the remarks about string arrays and ufuncs not being defined, there's a special case:
https://github.com/numpy/numpy/blob/26c1e04b3af572ed4ee1cc2f9ac3faca27858dc5/numpy/core/src/multiarray/arrayobject.c#L1281-L1309

?

**Comment by sklam** [comment]
Good find @stuartarchibald. So Numba should follow that and special case the handling for `==` for string types and not go down the ufunc route.

**Comment by stuartarchibald** [comment]
https://github.com/numba/numba/pull/7884 might help.

**Comment by dlee992** [comment]
I discussed with Seberg at this PR https://github.com/numpy/numpy/pull/21041, about unifying string ufuncs into the universal way. Still in discussion and need more NumPy PRs for `ufunc.types` stuff used in Numba `ufunc_find_matching_loop`. It's kind of needing long effort.
